rest-api/src/main/java/com/cloudbees/workflow/flownode/FlowNodeUtil.java

```diff
@@ -49,28 +71,113 @@
     private FlowNodeUtil() {
     }
 
-    public static long getNodeExecDuration(FlowNode node) {
-        List<FlowNode> childNodes = getChildNodes(node);
-        if (!childNodes.isEmpty()) {
-            long startTime = TimingAction.getStartTime(node);
-            long endTime = TimingAction.getStartTime(childNodes.get(childNodes.size() - 1));
+    public abstract static class CacheExtensionPoint implements ExtensionPoint {
+        public abstract Cache<String,List<FlowNode>> getExecutionCache();
+        public abstract Cache<String, RunExt> getRunCache();
+        public abstract Cache<FlowNode,String> getExecNodeNameCache();
+    }
 
-            if (startTime == 0) {
-                // The node is running and the time has not been marked on it yet.  Return 0 as the duration for now.
-                return 0;
+    // Used in testing where Jenkins is not running yet
+    private static final List<CacheExtension> FALLBACK_CACHES = Arrays.asList(new CacheExtension());
+
+    @Extension
+    public static class CacheExtension extends CacheExtensionPoint {
+
+        protected final Cache<String,List<FlowNode>> executionCache = CacheBuilder.newBuilder().weakValues().maximumSize(100).build();
```

Comment:
:ant: the use of `weakValues` with a `List` means that the cache is not very useful, since every GC cycle will throw away all the entries. The right approach is to use strong values but cache a `List<String>` IDs, where the actual `FlowNode`s are looked up on demand by ID; or to actually include all the metadata you need in some struct class in the value.